### ISSUE TITLE
Constructor fails due to ambiguous convert in self-referencing struct 

### ISSUE THREAD

**Issue by li1** [JuliaLang/julia#38567]
Hi, running the following code fails for me on
```
Julia Version 1.5.3
Commit 788b2c77c1 (2020-11-09 13:37 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
Environment:
  JULIA_PROJECT = /Users/malte/Documents/Dev/raicode
```
but also on 1.4, 1.3, and 1.6 nightly:

```
struct Inner{K,V} end
struct Passed{K,V} end

struct Outer{K,V}
    inner::Inner{K,V}
    nested_outer::Outer{Int,Int}

    function Outer{K,V}(::Passed{K,V}) where {K,V}
        return new{K,V}(Inner{K,V}())
    end
end

passed = Passed{Int,Int}()
outer = Outer{Int,Int}(passed)
```

It prints the following error message while executing the last line:
```
ERROR: LoadError: MethodError: convert(::Type{Inner{Int64,V}}, ::Inner{Int64,Int64}) is ambiguous. Candidates:
  convert(::Type{T}, x::T) where T>:Union{Missing, Nothing} in Base at missing.jl:68
  convert(::Type{T}, x::T) where T>:Nothing in Base at some.jl:33
  convert(::Type{T}, x::T) where T>:Missing in Base at missing.jl:67
  convert(::Type{T}, x) where T>:Nothing in Base at some.jl:34
  convert(::Type{T}, x::T) where T in Base at essentials.jl:171
  convert(::Type{T}, x) where T>:Missing in Base at missing.jl:69
  convert(::Type{T}, x) where T>:Union{Missing, Nothing} in Base at missing.jl:70
To resolve the ambiguity, try making one of the methods more specific, or adding a new method more specific than any of the existing applicable methods.
Stacktrace:
 [1] Outer{Int64,Int64}(::Passed{Int64,Int64}) at /Users/[...]/src/bug.jl:9
 [2] top-level scope at /Users/[...]/src/bug.jl:14
 [3] include(::String) at ./client.jl:457
 [4] top-level scope at REPL[1]:1
in expression starting at /Users/[...]/src/bug.jl:14
```

I expect it to create `Outer` without failing:
```
Outer{Int64,Int64}(Inner{Int64,Int64}(), #undef)
```

From the error message, it looks like the conversion is ambiguous, but I don't see why Julia shouldn't be able to figure out `V` for `Inner`.

Interestingly, the error does **not** occur if I change any of the following:
- only use one type argument (`K` instead of `{K,V}`)
- remove the self-referencing `nested_outer` field completely
- pick different types for `nested_outer` and `Passed (passed) / Outer (outer)`

For example, this would work:
```
struct Inner{K,V} end
struct Passed{K,V} end

struct Outer{K,V}
    inner::Inner{K,V}
    nested_outer::Outer{Int,Float64}

    function Outer{K,V}(::Passed{K,V}) where {K,V}
        return new{K,V}(Inner{K,V}())
    end
end

passed = Passed{Int,Int}()
outer = Outer{Int,Int}(passed)
```
Note that I changed `nested_outer`'s type to `Outer{Int,Float64}`, which doesn't match `Passed{Int,Int}`/`Outer{Int,Int}`.

Thanks for looking into it! ❤️ 

(also pinging @NHDaly who might be interested to follow along)

**Comment by NHDaly** [comment]
Based on the error message, it looks like it's constructing an invalid type for `Inner{Int64,V}`. `V` is not defined there, which I think is a bug.

```julia
julia> e = try
           outer = Outer{Int,Int}(passed)
       catch e
           e
       end
MethodError(convert, (Inner{Int64, V}, Inner{Int64, Int64}()), 0x0000000000007318)

julia> e.args[1]
Inner{Int64, V}

julia> Base.display_error(e)
ERROR: MethodError: ERROR: UndefVarError: T not defined
Stacktrace:
 [1] striptype(#unused#::Type{Inner{Int64, V}})
   @ Base ./errorshow.jl:333
 [2] show_convert_error(io::IOContext{Base.TTY}, ex::MethodError, arg_types_param::Any)
   @ Base ./errorshow.jl:210
 [3] showerror(io::IOContext{Base.TTY}, ex::MethodError)
   @ Base ./errorshow.jl:250
 [4] showerror(io::IOContext{Base.TTY}, ex::MethodError, bt::Nothing; backtrace::Bool)
   @ Base ./errorshow.jl:88
 [5] display_error(io::Base.TTY, er::MethodError, bt::Nothing)
   @ Base ./client.jl:98
 [6] display_error(er::MethodError, bt::Nothing)
   @ Base ./client.jl:108
```
I'm not sure how that got there, but it looks like a bug to me. :)

**Comment by NHDaly** [comment]
Also, for some more context, we ran into this in during development, and @li1 did a great job distilling it down to this MRE. But we were actually seeing it in our feature development.

Thanks for any help on this! 👍

**Comment by knuesel** [comment]
This seems to have been fixed, at least it works now on 1.8.2...

**Comment by KristofferC** [comment]
I think @vtjnash fixed this but I don't really know what to search for to find the PR number. Anyway, I think this can be closed.
